### ISSUE TITLE
Use the "persistent" IndexedDB option

### ISSUE THREAD

**Issue by kekkc** [greasemonkey/greasemonkey#2675]
Hi, 

after cleaning & deleting my "profile\storage\default" I realized that GM4.0 is currently saving the user scripts there. The usual extension storage "browser-extension-data\ID" just contains a storage.js with the content:
"{"previousWelcomeUrl":"https://www.greasespot.net/2017/09/greasemonkey-4-announcement.html"}"

Wouldn't it be better to use "browser-extension-data\ID" ffor all GM user extension data (incl. user scripts)? All my other extensions (adblock plus...) are using this folder, whereas "profile\storage\default" is usually just used by websites. This folder is a data security risk and will be automatically cleaned by FF:

https://www.heise.de/newsticker/meldung/Gefaehrdeter-Datenschutz-Firefox-loescht-lokale-Datenbanken-nicht-3835084.html
https://bugzilla.mozilla.org/show_bug.cgi?id=1047098

Greetz

**Comment by arantius** [comment]
> This folder is a data security risk

[citation needed]

**Comment by kekkc** [comment]
My fear is that my user scripts will be deleted if I clear my browsing history.

https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/storage 
Each extension has its own storage area, which can be split into different types of storage.

Although this API is similar to Window.localStorage it is **recommended that you don't use Window.localStorage in the extension code to store extension-related data**. **Firefox will clear data stored by extensions using the localStorage API in various scenarios** where users clear their browsing history and data for privacy reasons, while data saved using the storage.local API will be correctly persisted in these scenarios.

**Comment by Sxderp** [comment]
I don't know the folders but I do know the API points.
Greasemonkey stores userscripts in an `indexedDB`, which I'm going to assume is located in `profile\storage\default` as you have pointed out. Firefox will not _intentionally_ remove WebExtension data stored in `indexedDB`, actually I'm pretty sure Firefox will not _intentionally_ remove any WebExtension data which uses any of the storage WebExtension APIs.

> Although this API is similar to Window.localStorage it is recommended that you don't use Window.localStorage

`indexedDB` and `storage` are different from `window.localStorage`. What the message is saying is to not use `window.localStorage`. Greasemonkey does not use it.

> This folder is a data security risk

As for security risk, for extensions like Greasemonkey `storage.local` is actually more insecure. This is due to userscripts not being sandboxed from WebExtension permissions. A post I made, although a different issue, explains a bit about this. [Here](https://github.com/greasemonkey/greasemonkey/issues/2671#issuecomment-344612116). From what I remember, when I did my testing, indexedDB does not have that same problem.

**Comment by kekkc** [comment]
Thx for the explanation Sxderp. Interesting that there are advantages of using indexedDB.
However regarding security, I guess the sandbox problem will be fixed with https://bugzilla.mozilla.org/show_bug.cgi?id=1332273. Maybe then GM can use the regular Web Storage API?
```
defineLazyGetter(ProxyContextParent.prototype, "sandbox", function() {

  // NOTE: the required Blob and URL globals are used in the ext-registerContentScript.js

  // API module to convert JS and CSS data into blob URLs.

  return Cu.Sandbox(this.principal, {

    sandboxName: this.uri.spec,

    wantGlobalProperties: ["Blob", "URL"],

  });

});
```

**Comment by Sxderp** [comment]
> Maybe then GM can use the regular Web Storage API?

I still don't think so. `storage.local` is meant for smaller items. Like user config or preference settings. It is most definitely not for storing multi-kilobyte scripts.

**Comment by kekkc** [comment]
Sry, was too slow with my edit.

If indexedDB is a requirement I guess GM should create the storage as persistent at least:
var request = indexedDB.open("myDatabase", { version: 1, storage: "persistent" });

**Comment by Sxderp** [comment]
> If indexedDB is a requirement I guess GM should create the storage as persistent at least:

Not sure if it matters, but I don't see any reason not to add it. `indexedDb.open` is used in a couple locations. Change them and submit a pull request.
Do note, check that any existing scripts are still available when the change is made. If they're not there and the option key does matter we may have to look at this further.

**Comment by Sxderp** [comment]
>  arantius added this to the 4.x milestone 20 hours ago

Might want to reopen it?

**Comment by arantius** [comment]
Yep, thanks.

**Comment by Sxderp** [comment]
Turns out that if you add 'persistent' to the options FF treats it as a separate database. Now, assuming this is necessary (I can't find documentation that states whether it is or is not in WebExtensions) then all the scripts and stored values need to be copied.

**Comment by Sxderp** [comment]
I have a branch that migrates data over to the persistent database. Not sure if it's pushed. Can't remember, I'll check tomorrow.

**Comment by Sxderp** [comment]
I asked about this on the dev-addons mailing list.
> I have a question regarding the use of IndexedDB in WebExtensions. Is it required to pass `{storage: 'persistent'}` to IndexedDB.open in order to prevent Firefox from purging data, at some unknown point in the future, or are WebExtensions exempt from this? Assuming 'unlimitedStorage' is set in the manifest. I can't find any information specific to WebExtensions.

Response
> WebExtensions pages are not exempt from the IndexedDB "limits and eviction criteria" described in the following MDN doc page:
>
> https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API/Browser_storage_limits_and_eviction_criteria
>
> And so, the extension has to explicitly ask the storage to persist.
>
> On Firefox < 57 it was achieved by using the non-standard `{storage: 'persistent'}` option in the IndexedDB.open, but starting from Firefox >= 57 it can be also achieved with the storageManager API (which in my opinion is now preferable because it is the standardized way and the non-standard mode is likely to be removed in the future, and it can also turn an existent IndexedDB database into the persistent mode, without the need of migrating the data between IndexedDB db storages):
>
> https://developer.mozilla.org/en-US/docs/Web/API/StorageManager
>
> The 'unlimitedStorage' extension permission (as also described in the following section on MDN: https://developer.mozilla.org/en-US/Add-ons/WebExtensions/manifest.json/permissions#Unlimited_storage), doesn't enable the persist mode automatically on Firefox, but it allows the extension to enable the persistent storage mode at any time without prompting the user again (because the user has already accepted that the extension would be allowed to do it when the extension permissions have been accepted during the installation).
>
> As an additional side note, when an extension has the 'unlimitedStorage', it is also allowed to enable the persist mode from the background page, on the contrary the request would fail if it has to prompt the user and the extension page that is enabling the persistent mode is not a tab extension page.

Overview, looks like the database needs to be persisted but `{storage: 'persistent'}` should not be used. Instead the `StorageManager` API should. This will also make it easier to migrate as the data doesn't have to be manually moved from one DB instance to another.

**Comment by kekkc** [comment]
Thanks for getting the correct info, I find this totally confusing:

>https://developer.mozilla.org/en-US/docs/Web/API/StorageManager

This seems to be pointing to the new Storage API that is available with FF57 ( https://developer.mozilla.org/en-US/docs/Web/API/Storage_API ). And this Storage API just references IndexedDB databases, which are already used by GM. So it just comes down to the following?

1. Adjust manifest.json to include unlimited storage permission
`"permissions": [
  "*",
  "unlimitedStorage"
]`
2. Request persistent storage via StorageManager API (I guess in "src/bg/value-store.js" & "src/bg/user-script-registry.js")
`if (navigator.storage && navigator.storage.persist)  navigator.storage.persist().then(function(persistent){});`

Then there's no migration required and FF would simply copy the DB to the persistent folder? Would be great if that could be implemented to prevent FF from deleting anything in certain situations.

**Comment by Sxderp** [comment]
@kekkc , I did a test with similar code to what you had. And while `persistent` had a value of `true`, the files were still in the same location as discussed previously in this thread. Even through restarts of the browser.

I'm not sure if that is working as intended or what. Further, if you go to the `storage` tab of the dev-tools the Greasemonkey indexedDBs still have the `(default)` keyword. Tested on elementary OS (Ubuntu based, x64) FF 57 and FF 59 (nightly). Same results.

**Comment by kekkc** [comment]
>Tested on elementary OS (Ubuntu based, x64) FF 57 and FF 59 (nightly). Same results.

Seems then that they implemented the possibility to use the code, but the backend/box model is not yet implemented. I found:
https://bugzilla.mozilla.org/show_bug.cgi?id=1254428
https://bugzilla.mozilla.org/show_bug.cgi?id=1421690

Guess it makes no sense to change anything yet, because they'll implement the "cleaning rules" to all kind of storages.

**Comment by Sxderp** [comment]
Actually, looking at the bugs, I think this is fully implemented in regard to storing, at least, indexedDB data. And upon closer inspection of my profile I found a `permissions.sqlite` file. I dumped it and found the following:

```sql
INSERT INTO "moz_perms" VALUES(3,'moz-extension://a7382a42-007a-4648-8042-18b25ee6bdb0','persistent-storage',1,0,0,1514326961475);
```

That moz-extension ID happens to be the one for Greasemonkey on my test profile. So, I think it's working as intended in that the `permanent` folder is no longer being used.
